drivers/ipm/ipm_nrfx_ipc.c

```diff
@@ -0,0 +1,253 @@
+		}
+	}
+}
+
+static int ipm_nrf_send(struct device *dev, int wait, u32_t id,
+			const void *data, int size)
+{
+	if (id > NRFX_IPC_ID_MAX_VALUE) {
+		return -EINVAL;
+	}
+
+	if (size > 0) {
+		LOG_WRN("nRF driver does not support sending data over IPM");
+	}
+
+	gipm_send(id);
+	return 0;
+}
+
+static int ipm_nrf_max_data_size_get(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return 0;
+}
+
+static u32_t ipm_nrf_max_id_val_get(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return NRFX_IPC_ID_MAX_VALUE;
+}
+
+static void ipm_nrf_register_callback(struct device *dev,
+				      ipm_callback_t cb,
+				      void *context)
+{
+	nrfx_ipm_data.callback = cb;
+	nrfx_ipm_data.callback_ctx = context;
+}
+
+static int ipm_nrf_set_enabled(struct device *dev, int enable)
+{
+	/* Enable configured channels */
+	if (enable) {
+		irq_enable(DT_INST_0_NORDIC_NRF_IPC_IRQ_0);
+		nrfx_ipc_receive_event_group_enable((uint32_t)IPC_EVENT_BITS);
+	} else {
+		irq_disable(DT_INST_0_NORDIC_NRF_IPC_IRQ_0);
+		nrfx_ipc_receive_event_group_disable((uint32_t)IPC_EVENT_BITS);
+	}
+	return 0;
+}
+
+static int ipm_nrf_init(struct device *dev)
+{
+	gipm_init();
+	return 0;
+}
+
+static const struct ipm_driver_api ipm_nrf_driver_api = {
+	.send = ipm_nrf_send,
+	.register_callback = ipm_nrf_register_callback,
+	.max_data_size_get = ipm_nrf_max_data_size_get,
+	.max_id_val_get = ipm_nrf_max_id_val_get,
+	.set_enabled = ipm_nrf_set_enabled
+};
+
+DEVICE_AND_API_INIT(ipm_nrf, DT_INST_0_NORDIC_NRF_IPC_LABEL,
+		    ipm_nrf_init, NULL, NULL,
+		    PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
+		    &ipm_nrf_driver_api);
+
+#else
+
+struct vipm_nrf_data {
+	ipm_callback_t callback[NRFX_IPC_ID_MAX_VALUE];
+	void *callback_ctx[NRFX_IPC_ID_MAX_VALUE];
+	bool ipm_init;
+	struct device *ipm_device;
+};
+
+static struct vipm_nrf_data nrfx_vipm_data;
+
+static void vipm_dispatcher(uint32_t event_mask, void * p_context)
+{
+	while (event_mask)
+	{
+		uint8_t event_idx = __CLZ(__RBIT(event_mask));
+		event_mask &= ~(1UL << event_idx);
+		if(nrfx_vipm_data.callback[event_idx] != NULL) {
+			nrfx_vipm_data.callback[event_idx]
+				(nrfx_vipm_data.callback_ctx[event_idx],
+				 0,
+				 NULL);
+		}
+	}
+}
+
+static int vipm_nrf_max_data_size_get(struct device *dev)
+{
+	return ipm_max_data_size_get(dev);
+}
+
+static u32_t vipm_nrf_max_id_val_get(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return 0;
+}
+
+#define VIPM_DEVICE_1(_idx)									\
+static int vipm_nrf_##_idx##_init(struct device *dev)						\
+{												\
+	if(!nrfx_vipm_data.ipm_init) {								\
+		gipm_init();									\
+		nrfx_vipm_data.ipm_init = true;							\
+	}											\
+	return 0;										\
+}												\
+												\
+static int vipm_nrf_##_idx##_send(struct device *dev, int wait, u32_t id,			\
+				  const void *data, int size)					\
+{												\
+	if (!IS_ENABLED(CONFIG_IPM_MSG_CH_##_idx##_ENABLE)) {					\
+		LOG_ERR("IPM_" #_idx " is not enabled");					\
+		return -EINVAL;									\
```

Comment:
Isn't this code-path impossible due to https://github.com/zephyrproject-rtos/zephyr/pull/20467/files#diff-ec888dfc90d8c3eb9142711f8e6a3281R226 ?

Reply:
Right, removed